tsdb/index/index.go

```diff
@@ -1511,6 +1511,51 @@ func (r *Reader) LabelValues(name string, matchers ...*labels.Matcher) ([]string
 	return values, nil
 }
 
+// LabelNamesFor returns all the label names for the series referred to by IDs.
+// The names returned are sorted.
+func (r *Reader) LabelNamesFor(ids ...uint64) ([]string, error) {
+	// Gather offsetsMap the name offsetsMap in the symbol table first
+	offsetsMap := make(map[uint32]struct{})
+	for _, id := range ids {
+		offset := id
+		// In version 2 series IDs are no longer exact references but series are 16-byte padded
+		// and the ID is the multiple of 16 of the actual position.
+		if r.version == FormatV2 {
+			offset = id * 16
+		}
+
+		d := encoding.NewDecbufUvarintAt(r.b, int(offset), castagnoliTable)
+		buf := d.Get()
+		if d.Err() != nil {
+			return nil, errors.Wrap(d.Err(), "label names for")
+		}
+
+		offsets, err := r.dec.LabelNamesOffsetsFor(buf)
+		if err != nil {
+			return nil, storage.ErrNotFound
+		}
+		for _, off := range offsets {
```

Comment:
I was thinking first to have an `[][]uint32 allOffsets` with all offsets slices here and then make an k-way merge using a heap. However, that would mean that allOffsets could potentially be the entire copy of the all the series with their symbols offsets, which seemed unnecessary so I decided to go with a map and keep deduplicating them as we get them.